### ISSUE TITLE
Node 0.12 stable fails on strict mode

### ISSUE THREAD

**Issue by tresdin** [joyent/node#9187]
I found this error when I run node with `node --use_strict myapp.js`. What can I do in this case?

![syntaxerroronstrictmode](https://cloud.githubusercontent.com/assets/7781091/6142046/b074693e-b1e1-11e4-82d1-c22ddbab9963.png)

**Comment by misterdjules** [comment]
If you need to load the `crypto` module and to pass the `--use_strict` flag to V8, I don't think there is anything you can do with node v0.12.0 to prevent this error.

May I ask why you need to pass `--use_strict`? 

Adding this issue to the 0.12.1 milestone.

**Comment by tresdin** [comment]
@misterdjules Just because it's more convenient than inserting `use strict` string on top of each file. BTW, this flag `--use_strict` is pretty useless since several internal and external modules fail on strict mode. Node can fix its modules, but the external one. I really think Node should provide some standards, which may be similar to `ECMAScript`. Otherwise, sooner or later, those external guys will mess it up.

**Comment by OrangeDog** [comment]
See also #7479.

**Comment by clintdw** [comment]
Maybe I don't fully comprehend what the `--use_strict` flag to V8 (via node) is intended for (internal testing for the V8 team versus forcing correctness for Javascript developers), but when I first started using node, I thought that it existed to serve the same purpose as it existed in Perl's `use strict` [1]: to enforce correct behavior and warn (stop) you when you're being sloppy with your code to the point where your code isn't doing what you think it is. It would have been a great blunder if a production release of Perl had bundled modules that didn't passed the strictness test. Likewise, anyone contributing a third party module on CPAN was also expected to write `use strict` compliant code, or at least any module writer who wanted to be taken seriously.

If V8's `--use_strict` isn't what node developers want to use for this purpose, I certainly hope that something similar to perl's `use strict` is created by the node team to serve a similar purpose. While you're at it, copying `perl -cw` to validate a program without actually running it would be useful in node as well.

[1] http://perldoc.perl.org/strict.html

**Comment by rlidwka** [comment]
> If V8's --use_strict isn't what node developers want to use for this purpose

No, `--use_strict` is supposed to be used internally in V8 tests. See https://github.com/joyent/node/issues/7479.

If you want to get warnings about "sloppy" code, you can use linters like `jshint`, `eslint`, etc.

**Comment by misterdjules** [comment]
@tresdin Thanks for the information. Like several others pointed out already here and in #7479 , passing `--use-strict` to node, or more generally enabling `"use strict";` for all code loaded in a given runtime is not necessarily a good idea, since not all code available complies to the rules of the strict mode.

Nevertheless, it seems that we can't really deprecate `--use-strict`, at least for the short term future. Always breaking users who use this flag would also generate a lot of questions and confusion. Fixing the code that breaks seems to be more effective than pointing to a FAQ or some documentation. For these reasons, I put together [a PR that fixes some of node's built-in modules that currently fail to load with `--use-strict`](https://github.com/joyent/node/pull/9237).

Actually, if we decide to go down that path, I'm wondering if we should also add `"use strict";` in every module file.

**Comment by clintdw** [comment]
> I put together a PR that fixes some of node's built-in modules that currently fail to load with --use-strict.

Excellent.

> Actually, if we decide to go down that path, I'm wondering if we should also add "use strict"; in every module file.

@tresdin this sounds like a great idea, otherwise variations of bugs like this will appear wherever a built-in module isn't compliant.

**Comment by cjihrig** [comment]
@misterdjules your PR can't be merged. What if we included `'use strict';` at the top of all the core files? It worked fine in [io.js](https://github.com/iojs/io.js/commit/21130c7d6fcaee666f33735768c060be2e06614a).

**Comment by misterdjules** [comment]
@cjihrig What do you mean by "can't be merged"? Yes, adding `'use strict';` at the top of all core files is also something that I suggested. It would not necessarily remove the need for a test though.

**Comment by cjihrig** [comment]
Sorry, I just meant that there are merge conflicts.

**Comment by misterdjules** [comment]
@cjihrig It applies cleanly with ` curl https://github.com/joyent/node/pull/9237.patch | git am -`:

```
➜  v0.12 git:(v0.12) ✗ git remote -v | grep upstream
upstream	git@github.com:joyent/node.git (fetch)
upstream	git@github.com:joyent/node.git (push)
➜  v0.12 git:(v0.12) ✗ git fetch upstream
➜  v0.12 git:(v0.12) ✗ git rev-parse --short HEAD
b3aa876
➜  v0.12 git:(v0.12) ✗ git rev-parse --short upstream/v0.12
b3aa876
➜  v0.12 git:(v0.12) ✗ curl https://github.com/joyent/node/pull/9237.patch | git am -
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7223    0  7223    0     0   6367      0 --:--:--  0:00:01 --:--:--  6369
Applying: src: fix builtin modules failing with --use-strict
➜  v0.12 git:(v0.12) ✗
```

Am I missing something?

**Comment by cjihrig** [comment]
@misterdjules I meant the grey "We can’t automatically merge this pull request." on https://github.com/joyent/node/pull/9237. Is it showing up green for you?

**Comment by misterdjules** [comment]
@cjihrig Thank you, and sorry for not realizing that earlier :) I usually do not check the "merge section" in any PR because we don't use the merge button. I don't know why GitHub says it cannot be merged, I'll investigate further.

**Comment by tresdin** [comment]
@misterdjules  I think `--use-strict` should be globally applied by default. And Node need to provide a `--non-strict` flag instead. Don't be afraid of this decision. ECMAScript guys really want to do the same thing, but they can't. Just because it's too late. This is not true with the 5 years old Nodejs.

**Comment by rlidwka** [comment]
>  I think --use-strict should be globally applied by default. And Node need to provide a --non-strict flag instead. Don't be afraid of this decision.

Sure. If we add a way to opt-out of it for each individual module (i.e. `"use no strict"` flag). But until then strict mode should never be the default.

**Comment by misterdjules** [comment]
Fixed in b233131, thank you!

**Comment by misterdjules** [comment]
Also, a follow-up commit that enables strict mode in all built-in modules has landed in master with https://github.com/joyent/node/commit/ef4344311e19a4f73c031508252b21712b22fe8a.
